ASoC: SOF: add ext manifest 4 firmware support

bardliao opened the pull request:
SOF can support a new firmware extension manifest from ABI version 4. This
patch adds extension manifest 4 parser.

Signed-off-by: Rander Wang <rander.wang@intel.com>

@plbossart @ujfalusi I am not sure can we use bitfields in extension manifest struct. I will modify it if it is not suitable.

RanderWang commented on include/sound/sof/ext_manifest4.h:
```diff
@@ -0,0 +1,135 @@
+#define SOF_MAN4_FW_HDR_OFFSET	0x2000
+
+/*********************************************************************
+ *     css_manifest hdr
```
there is a extended manifest before css hdr, please help to add it.

bardliao replied:
Done

ujfalusi commented on include/sound/sof/ext_manifest4.h:
```diff
@@ -0,0 +1,135 @@
+	uint32_t domain_dp : 1; /* support DP domain */
+	uint32_t lib_code : 1; /* determines if module is place holder for common library code */
+	uint32_t _rsvd : 24;
+} __packed;
```
It might be only me, but parsing the bit definition like this is not too user and compiler friendly.
With gcc the `struct sof_man4_module_type` is four bytes (uint32), and not sure what the __packed supposed to do.
For a quick test:
```c
#include <stdio.h>

typedef unsigned int uint32_t;

struct sof_man4_module_type_1 {
	uint32_t load_type; /* MT_BUILTIN, MT_LOADABLE */
	uint32_t auto_start; /* 0 - manually created, 1 - created by Module Manager */
	uint32_t domain_ll; /* support LL domain */
	uint32_t domain_dp; /* support DP domain */
	uint32_t lib_code; /* determines if module is place holder for common library code */
	uint32_t _rsvd;
};

struct sof_man4_module_type_2 {
	uint32_t load_type : 4; /* MT_BUILTIN, MT_LOADABLE */
	uint32_t auto_start : 1; /* 0 - manually created, 1 - created by Module Manager */
	uint32_t domain_ll : 1; /* support LL domain */
	uint32_t domain_dp : 1; /* support DP domain */
	uint32_t lib_code : 1; /* determines if module is place holder for common library code */
	uint32_t _rsvd : 24;
};

struct sof_man4_module_type_3 {
	uint32_t load_type : 4; /* MT_BUILTIN, MT_LOADABLE */
	uint32_t auto_start : 1; /* 0 - manually created, 1 - created by Module Manager */
	uint32_t domain_ll : 1; /* support LL domain */
	uint32_t domain_dp : 1; /* support DP domain */
	uint32_t lib_code : 1; /* determines if module is place holder for common library code */
	uint32_t _rsvd : 24;
} __packed;

int main(int argc, char *argv[])
{
	struct sof_man4_module_type_1 type_1;
	struct sof_man4_module_type_2 type_2;
	struct sof_man4_module_type_3 type_3;

	printf( "sizeof sof_man4_module_type_1 : %lu\n", sizeof(type_1));
	printf( "sizeof sof_man4_module_type_2 : %lu\n", sizeof(type_2));
	printf( "sizeof sof_man4_module_type_3 : %lu\n", sizeof(type_3));

	return 0;
}
```
Results:
```
sizeof sof_man4_module_type_1 : 24
sizeof sof_man4_module_type_2 : 4
sizeof sof_man4_module_type_3 : 4
```

bardliao replied:
@ujfalusi The bit definition allocation is from firmware side. We can only use either `struct sof_man4_module_type` or `uint32_t sof_man4_module_type`. And it is easier to access the `load_type` field by `type.load_type` than using a macro to me.
Regarding `__packed`, it is to avoid the structures to be padded. There is no different between your `struct sof_man4_module_type_2` and `struct sof_man4_module_type_3` is because there is only one item in the structure. You can see the size of sof_man4_module_type_2 is 8 and sof_man4_module_type_3 is 5 if you add a `char` item in the structure.
BTW, I have to use ` __attribute__((__packed__))` instead of `__packed` when I tested it. But `__packed` is preferred in kernel space.

ujfalusi replied:
@bardliao, right, but consider if the compiler for a given platform is handling the C bitfield in a different way (or not at all).
Also if you see a hex dump of `0x00000053`, how would you quickly check what it is?
```c
struct sof_man4_module_type {
	uint32_t load_type : 4; /* MT_BUILTIN, MT_LOADABLE */
	uint32_t auto_start : 1; /* 0 - manually created, 1 - created by Module Manager */
	uint32_t domain_ll : 1; /* support LL domain */
	uint32_t domain_dp : 1; /* support DP domain */
	uint32_t lib_code : 1; /* determines if module is place holder for common library code */
	uint32_t _rsvd : 24;
} __packed;
```
VS
```c
#define SOF_MAN4_MODULE_LOAD_TYPE	GENMASK(0, 3)
#define SOF_MAN4_MODULE_AUTO_START 	BIT(4)
#define SOF_MAN4_MODULE_DOMAIN_LL	BIT(5)
#define SOF_MAN4_MODULE_DOMAIN_DP	BIT(6)
#define SOF_MAN4_MODULE_LIB_CODE	BIT(7)
```
If we also want to debug this at some point then probably do something like:
```c
union sof_man4_module_type {
	uint32_t as_uint32;
	struct {
		uint32_t load_type : 4; /* MT_BUILTIN, MT_LOADABLE */
		uint32_t auto_start : 1; /* 0 - manually created, 1 - created by Module Manager */
		uint32_t domain_ll : 1; /* support LL domain */
		uint32_t domain_dp : 1; /* support DP domain */
		uint32_t lib_code : 1; /* determines if module is place holder for common library code */
		uint32_t _rsvd : 24;
	} __packed;
};
```

bardliao replied:
@ujfalusi You convinced me, I will use macros. Thanks.

plbossart commented on include/sound/sof/ext_manifest4.h:
```diff
@@ -0,0 +1,135 @@
+	uint32_t domain_ll : 1; /* support LL domain */
+	uint32_t domain_dp : 1; /* support DP domain */
+	uint32_t lib_code : 1; /* determines if module is place holder for common library code */
+	uint32_t _rsvd : 24;
```
does the kernel need to know all this? it doesn't look like it's actually needed or processed?

RanderWang replied:
yes, we need these flag to calculate memory usage of each module. The related patches will be submitted later  by bard.

plbossart commented on include/sound/sof/ext_manifest4.h:
```diff
@@ -0,0 +1,135 @@
+struct sof_man4_fw_img_flags {
+	uint32_t _rsvd0 : 31;
+	uint32_t tp : 1;
+} __packed;
```
do we need to know this in the driver?

plbossart commented on include/sound/sof/ext_manifest4.h:
```diff
@@ -0,0 +1,135 @@
+	/* This part may change to contain any additional data for BaseFw that is skipped by ROM */
+	uint32_t		hw_buf_base_addr;
+	uint32_t		hw_buf_length;
+	uint32_t		load_offset; /* This value is used by ROM */
```
same, do we need to know all this?

bardliao replied:
We may not need to know every elements of `struct sof_man4_fw_binary_header {}`, but we can't get the elements we need at the correct position if we remove some of them, right?

plbossart commented on include/sound/sof/ext_manifest4.h:
```diff
@@ -0,0 +1,135 @@
+	uint32_t	type : 4;
+	uint32_t	_rsvd1 : 4;
+
+	uint32_t	length : 16; /* segment length in pages */
```
does the kernel need to know? That looks more of a uapi between userpace (rimage) and firmware.

plbossart reviewed (commented):
Confused on what parts are actually needed by the kernel.

ranj063 commented on sound/soc/sof/loader.c:
```diff
@@ -358,6 +359,91 @@ static int snd_sof_fw_ext_man_parse(struct snd_sof_dev *sdev,
+	if (hdr->id == SOF_EXT_MAN4_MAGIC_NUMBER) {
+		fw_offset = hdr->len;
+	} else {
+		dev_err(sdev->dev, "invalid manifest 4 FW");
```
how will we ever see this? we wont call snd_sof_fw_ext_man4_parse unless the condition is true no?

bardliao replied:
> how will we ever see this? we wont call snd_sof_fw_ext_man4_parse unless the condition is true no?

Right, thanks for pointing this out.

ranj063 commented on sound/soc/sof/loader.c:
```diff
@@ -358,6 +359,91 @@ static int snd_sof_fw_ext_man_parse(struct snd_sof_dev *sdev,
+		return -ENOMEM;
+
+	sdev->fw_module_num = fw_header->num_module_entries;
+	sdev->fw_modules = module_entry;
```
it will be more intuitive if we do sdev->fw_modules = devm_kmalloc_array() and then module_entry = sdev->fw_modules no?

bardliao replied:
> it will be more intuitive if we do sdev->fw_modules = devm_kmalloc_array() and then module_entry = sdev->fw_modules no?

Yes, will edit.

bardliao commented:
> Confused on what parts are actually needed by the kernel.

Thanks @plbossart. I checked those elements and we don't need most of them.

ranj063 commented on include/sound/sof/ext_manifest4.h:
```diff
@@ -0,0 +1,108 @@
+	uint32_t	cps;		/* cycles per second */
+	uint32_t	ibs;		/* input buffer size (bytes) */
+	uint32_t	obs;		/* output buffer size (bytes) */
+	uint32_t	module_flags;	/* flags, res for future use */
```
can we expand res to reserved so that readers dont have to guess what it stands for?

ujfalusi replied:
I love the `cps`, `ibs` and `obs` along with `cpc` also, `par` is descriptive as well.

ranj063 commented on include/sound/sof/ext_manifest4.h:
```diff
@@ -0,0 +1,108 @@
+	uint32_t	obs;		/* output buffer size (bytes) */
+	uint32_t	module_flags;	/* flags, res for future use */
+	uint32_t	cpc;		/* cycles per single run */
+	uint32_t	obls;		/* output block size, res for future use */
```
same here

RanderWang replied:
"for future use" I have no idea

ranj063 commented on include/sound/sof/ext_manifest4.h:
```diff
@@ -0,0 +1,108 @@
+	uint32_t	ibs;		/* input buffer size (bytes) */
+	uint32_t	obs;		/* output buffer size (bytes) */
+	uint32_t	module_flags;	/* flags, res for future use */
+	uint32_t	cpc;		/* cycles per single run */
```
single run of what? sorry the comment is quite vague, could you please elaborate?

bardliao replied:
> single run of what? sorry the comment is quite vague, could you please elaborate?

@RanderWang Could you comment on this?

RanderWang replied:
it is related to schedule. In each schedule period, each module is scheduled to process a batch of sample data, generally 1ms data. Here single run means the time span that each module used.

plbossart replied:
then 'max cycles per scheduling period'?

bardliao replied:
> then 'max cycles per scheduling period'?

Updated, thanks

ranj063 commented on sound/soc/sof/loader.c:
```diff
@@ -358,6 +359,86 @@ static int snd_sof_fw_ext_man_parse(struct snd_sof_dev *sdev,
+
+		memcpy(&module_entry->type, &fm_entry->type, sizeof(fm_entry->type));
+
+		/* bringup fw starts at zero */
```
why is this comment relevant here?

RanderWang replied:
bringup fw is not a common module but it is defined module config.

ranj063 replied:
ok but still not getting why this has to be mentioned right at this line

ranj063 commented on sound/soc/sof/loader.c:
```diff
@@ -358,6 +359,86 @@ static int snd_sof_fw_ext_man_parse(struct snd_sof_dev *sdev,
+		return -ENOMEM;
+
+	sdev->fw_module_num = fw_header->num_module_entries;
+	module_entry = sdev->fw_modules;
```
wow the variable names and struct names are so confusing. You  have variable fw_entry pointing to a struct named sof_man4_module_entry. And you have a variable module_entry pointing to a struct name sof_ipc4_fw_modules. I dont know how to decipher this. I trust this is doing the right thing but so hard to read and understand what these are used for.

bardliao replied:
> wow the variable names and struct names are so confusing. You have variable fw_entry pointing to a struct named sof_man4_module_entry. And you have a variable module_entry pointing to a struct name sof_ipc4_fw_modules. I dont know how to decipher this. I trust this is doing the right thing but so hard to read and understand what these are used for.

fw_entry is pointing to the data of firmware binary, and we copy some items to module_entry (sdev->fw_modules) which will be used by SOF. @RanderWang Please add your comments if I miss something.

RanderWang replied:
yes, Here we don't exactly copy the module config in FW manifest, but create another structure to make it more simple and general  , then copy some of them. I also discussed with bard about this. @ranj063 do you prefer to fully save module config ?

ujfalusi replied:
Can we collect the `4` related new pointers, variables under some struct and not spread it around in sdev?

bardliao replied:
> Can we collect the `4` related new pointers, variables under some struct and not spread it around in sdev?

I don't know if we want to go with this. It will make `4` so special. What do you think? @lgirdwood @RanderWang @plbossart @ranj063 ?

ujfalusi replied:
Well, it _is_ special ;)
If it is just two new member then I guess it is fine. I can not recall if we have similar 'global' storage for similar things, if we do then it would make things cleaner to stuff them aside

ranj063 replied:
there are 2 separate issues. I was literally only talking about making the variable names more intuitive for readers to understand whats going on. But on the other issue of whether ew should copy and save only certain modules depends on how scalable that is. Why do we do this today? Is it because SOF doesnt support all modules? What happens if and when we support all modules? Will we still need to do this?

bardliao replied:
> Well, it _is_ special ;) If it is just two new member then I guess it is fine. I can not recall if we have similar 'global' storage for similar things, if we do then it would make things cleaner to stuff them aside

I think the two member below are the only new member specific to `4`. So I guess it is fine. In other words, I will not add new 4 specific members in the future,
```
        int fw_module_num;
        struct sof_ipc4_fw_modules *fw_modules;
```

bardliao replied:
> there are 2 separate issues. I was literally only talking about making the variable names more intuitive for readers to understand whats going on. But on the other issue of whether ew should copy and save only certain modules depends on how scalable that is. Why do we do this today? Is it because SOF doesnt support all modules? What happens if and when we support all modules? Will we still need to do this?

Currently, we only store what we really need in `struct sof_ipc4_fw_modules {}`, but I don't know if we will need more in the future. Maybe we can copy the entire `sof_man4_module_entry` so we don't need to add new variables in the future

ujfalusi commented on sound/soc/sof/loader.c:
```diff
@@ -358,6 +359,86 @@ static int snd_sof_fw_ext_man_parse(struct snd_sof_dev *sdev,
+	const struct sof_ext_manifest4_hdr *hdr;
+
+	hdr = (struct sof_ext_manifest4_hdr *)fw->data;
+	if (hdr->id == SOF_EXT_MAN4_MAGIC_NUMBER) {
```
`hdr->id` is `magic`, right?

bardliao replied:
> `hdr->id` is `magic`, right?

Yes, hdr->id is a magic number

ujfalusi commented on sound/soc/sof/loader.c:
```diff
@@ -358,6 +359,86 @@ static int snd_sof_fw_ext_man_parse(struct snd_sof_dev *sdev,
+
+	hdr = (struct sof_ext_manifest4_hdr *)fw->data;
+	if (hdr->id == SOF_EXT_MAN4_MAGIC_NUMBER) {
+		dev_dbg(sdev->dev, "%s manifest4 fw is detected\n", __func__);
```
Why not print if manifest3 is detected? Why print at all?

bardliao replied:
> Why not print if manifest3 is detected? Why print at all?

Will remove it.

ujfalusi commented on include/sound/sof/ext_manifest4.h:
```diff
@@ -0,0 +1,108 @@
+
+struct sof_ext_manifest4_hdr {
+	u32 id;
+	u32 len;
```
Length of what?

bardliao replied:
> Length of what?

length of extension manifest. I will add a comment here.

ujfalusi commented on sound/soc/sof/loader.c:
```diff
@@ -358,6 +359,86 @@ static int snd_sof_fw_ext_man_parse(struct snd_sof_dev *sdev,
+	fw_offset = hdr->len;
+
+	fw_header = (struct sof_man4_fw_binary_header *)(fw->data + fw_offset +
+							 SOF_MAN4_FW_HDR_OFFSET);
```
we have so many offsets that it is hard to follow, can you use `hdr->len` instead of `fw_offset`, it is only used here and then returned at the end.
between `hdr->len` and `SOF_MAN4_FW_HDR_OFFSET` we have grey area? 0x2000 is a lot of wasted space.

bardliao replied:
> we have so many offsets that it is hard to follow, can you use `hdr->len` instead of `fw_offset`, it is only used here and then returned at the end. 

Will use `hdr->len` instead of `fw_offset`.

> between `hdr->len` and `SOF_MAN4_FW_HDR_OFFSET` we have grey area? 0x2000 is a lot of wasted space.

I checked the firmware binary, there are some data followed by a lot of 0s in the SOF_MAN4_FW_HDR_OFFSET area. I believe that area is used by some struct.

marc-hb replied:
This `0x2000` has been a constant for a long time: https://github.com/thesofproject/sof/commit/8885b465df53

To test https://github.com/thesofproject/sof/pull/4889 I ran the latest `tools/sof_ri_info/sof_ri_info.py` on all platform binaries since v1.7 and it parsed all of them. Example:

```
Reading SOF ri image sof-bin/v1.9.x/sof-v1.9/intel-signed/sof-tgl-h.ri
    File size 0x68300 (426752)
    0x0     Looking for Extended Manifest
    0x0     Extended Manifest (XMan)
    0x300   Looking for CSE Manifest
    0x300   CSE Manifest ($CPD)
    0x308   # of entries 3
    0x314   Reading CSE Manifest entry 0...
    0x32c   ... CSE Entry name ADSP.man file offset 0x35c length 1124
    ...
    0x2300  ADSP Manifest ($AM1)
    0x2334  Module Entry signature found ($AME)
    0x23a8  Module Entry signature found ($AME)
    Parsing finished

    SOF Binary sof-bin/v1.9.x/sof-v1.9/intel-signed/sof-tgl-h.ri size 0x68300

      Extended Manifest ver 1.0.0 length 768

      CSE Manifest ver 0x102 checksum 0x0 partition name ADSP

        ADSP.man (CSS Manifest) type 0x4 file offset 0x35c hdr_len 900 ver 0x21000 date 2021/09/23
          Rsvd0 0x1
          Modulus size (dwords) 96
            d3 72 92 99 4e b9 cd 67 ... 5d 62 4a 80 96 31 f8 b5 (TGL Intel prod key)
          Exponent size (dwords) 1
      ....
      ....

      cavs0015 (ADSP Manifest) file offset 0x2300 name ADSPFW build ver 1.9.0.1 feature mask 0xffff image flags 0x0
        HW buffers base address 0x0 length 0x0
        Load offset 0x30000


```

ujfalusi commented on sound/soc/sof/loader.c:
```diff
@@ -358,6 +359,86 @@ static int snd_sof_fw_ext_man_parse(struct snd_sof_dev *sdev,
+		if (fm_entry->cfg_count)
+			module_entry->bss_size = fm_config[fm_entry->cfg_offset].is_bytes;
+
+		memcpy(&module_entry->type, &fm_entry->type, sizeof(fm_entry->type));
```
`module_entry->type = fm_entry->type;` ?

bardliao replied:
> `module_entry->type = fm_entry->type;` ?

ack

ujfalusi commented on sound/soc/sof/sof-priv.h:
```diff
@@ -56,6 +56,8 @@ extern int sof_core_debug;
 #define SOF_IPC_DSP_REPLY		0
 #define SOF_IPC_HOST_REPLY		1
 
+#define SOF_MODULE_NAME_LEN		16
```
`MAX_MODULE_NAME_LEN` is 8 and you memcpy it from fw to local struct

bardliao replied:
> `MAX_MODULE_NAME_LEN` is 8 and you memcpy it from fw to local struct

Nice catch. Will change it to 8

ujfalusi commented on include/sound/sof/ext_manifest4.h:
```diff
@@ -0,0 +1,108 @@
+struct sof_man4_module_entry {
+	uint32_t	id;
+	uint8_t		name[MAX_MODULE_NAME_LEN];
+	uint8_t		uuid[16];
```
`UUID_SIZE` is 16?

bardliao replied:
> `UUID_SIZE` is 16?

Yes, I will include <linux/uuid.h> and use `UUID_SIZE`. Thanks.

ujfalusi commented on sound/soc/sof/loader.c:
```diff
@@ -358,6 +359,86 @@ static int snd_sof_fw_ext_man_parse(struct snd_sof_dev *sdev,
+
+		/* align to dw */
+		dw_count = (module_entry->instance_max_count + 31) >> 5;
+		module_entry->instance_id = devm_kzalloc(sdev->dev, dw_count, GFP_KERNEL);
```
are you sure about this?
if `instance_max_count` is **27**, then `dw_count` will be **1**, so one byte for a bit array which is somehow connected to `instance_max_count`

You might wanted something like:
```c
module_entry->instance_id = devm_kmalloc_array(dev, BITS_TO_U32(module_entry->instance_max_count),
					       sizeof(u32), GFP_KERNEL);
```
But I would go with `BITS_TO_LONGS`

RanderWang replied:
good catch!  you got my idea.

bardliao replied:
Thanks @ujfalusi I would like to use BITS_TO_U32 since module_entry->instance_id is *u32. May I know why you prefer BITS_TO_LONGS?

kv2019i commented on include/sound/sof/ext_manifest4.h:
```diff
@@ -0,0 +1,108 @@
+	/* This part must be unchanged to be backward compatible with SPT-LP ROM */
+	uint32_t	id;
+	uint32_t	len; /* sizeof(sof_man4_fw_binary_header) in bytes */
+	uint8_t		name[MAX_FW_BINARY_NAME];
```
Here's some odd whitespace between "uint8_t" and "name"

kv2019i commented on include/sound/sof/ext_manifest4.h:
```diff
@@ -0,0 +1,108 @@
+	uint8_t		name[MAX_MODULE_NAME_LEN];
+	uint8_t		uuid[16];
+	uint32_t	type;
+	uint8_t		hash[DEFAULT_HASH_SHA256_LEN];
```
The added whitespace here as well.

bardliao replied:
> The added whitespace here as well.

It is intentional, they are 2 tabs not 1 whitespace + 1 tab. I added them to align the variable fields.

kv2019i replied:
Ack @bardliao

kv2019i reviewed (changes requested):
If I compare this to include/sound/sof/ext_manifest.h (for v3), this is quite specific (for example many fields for ROM only usage). I wonder if this should be hidden as a vendor manifest instead, e.g. Intel specific manifest.

If the amount of data SOF driver needs to get from manifest stays low, we could have a function interface to get data (like FW version and max IPC MSG size) and this could be coded in a vendor specific manifest structure.

Of course, if SOF core needs to get more data for IPC4 (like the module description data), then this become more complex. At minimum in this case, we should have clear documentation which of the manifest fields are relevant to driver and describe them better (like in include/sound/sof/ext_manifest.h -> all fields documented).

bardliao commented:
changes: 
1. Change type of `*fw_modules` from `struct sof_ipc4_fw_modules {}` to `void` in `struct snd_sof_dev {}` to let `*fw_modules` be more general.
2. Use `struct sof_man4_module_entry` directly in `struct sof_ipc4_fw_modules {}`.

@ujfalusi @ranj063 Does it look a bit better now? I am also trying to rename the variables in snd_sof_fw_ext_man4_parse(), but I don't have a good idea so far. Do you have any suggestion?

ranj063 commented on include/sound/sof/ext_manifest4.h:
```diff
@@ -0,0 +1,110 @@
+	uint32_t	file_offset;
+} __packed;
+
+struct sof_man4_module_entry {
```
why do we call this struct as sof_man4_module_entry? WHat does the entry signify here? Why not call it simply "struct sof_man4_module"

ranj063 commented on sound/soc/sof/loader.c:
```diff
@@ -358,6 +360,75 @@ static int snd_sof_fw_ext_man_parse(struct snd_sof_dev *sdev,
+	dev_dbg(sdev->dev, " fw %s: header length %x, module num %d", fw_header->name,
+		fw_header->len, fw_header->num_module_entries);
+
+	fm_entry = (struct sof_man4_module_entry *)((void *)fw_header + fw_header->len);
```
if you change the name of struct from module_entry to just module, then fm_entry for this variable makes more sense. It points to the first module or the entry module.

ranj063 commented on sound/soc/sof/loader.c:
```diff
@@ -358,6 +360,75 @@ static int snd_sof_fw_ext_man_parse(struct snd_sof_dev *sdev,
+		fw_header->len, fw_header->num_module_entries);
+
+	fm_entry = (struct sof_man4_module_entry *)((void *)fw_header + fw_header->len);
+	fm_config = (struct sof_man4_module_config *)(fm_entry + fw_header->num_module_entries);
```
if I look at the breakdown of the manifest above, after all the modules you have "FW Content". It doesnt seem to se show module config. Am I reading this wrong?

bardliao replied:
@ranj063 module config is used at line 402 `module_entry->bss_size = fm_config[fm_entry->cfg_offset].is_bytes;`. Does it answer your question?

ranj063 replied:
No @bardliao  what I was asking was that on line 45 above, after module [n] , it shows fw content. But in the line above fw_config Is pointing to that isn't i

bardliao replied:
> No @bardliao what I was asking was that on line 45 above, after module [n] , it shows fw content. But in the line above fw_config Is pointing to that isn't i

I got it now. Fixed.

ranj063 commented on sound/soc/sof/loader.c:
```diff
@@ -358,6 +360,75 @@ static int snd_sof_fw_ext_man_parse(struct snd_sof_dev *sdev,
+		module_entry->man4_module_entry.id = i;
+
+		/* align to dw */
+		module_entry->instance_id =
```
what is the purpose of the instance ID? It doesnt look like it is incremental per module.

bardliao replied:
> what is the purpose of the instance ID? It doesnt look like it is incremental per module.

The `instance_id` is to map modules and SOF widget. A module will map to a SOF widget and they will have the same id. @RanderWang Please correct me if I was wrong.

ranj063 replied:
Ok 2 questions then,.why is it an array? Can we drop it for now and add it later when you show how it is used?

bardliao replied:
> Ok 2 questions then,.why is it an array? Can we drop it for now and add it later when you show how it is used?

Sorry, "The instance_id is to map modules and SOF widget. A module will map to a SOF widget and they will have the same id." is not precise. `instance_id` here is a bit field. Each bit represents an active swidget and the amount of swidget (`instance_max_count`) is dynamic. `instance_max_count` could be 5 or 50 for example. That's why we use n dynamic size array. We could use `u64 instance_id;` if `instance_max_count` is guarantee less than 64. 
Yes, I can drop it for now and add it later.

ranj063 reviewed (dismissed):
looks good. Thanks for being patient with the reviews @bardliao

kv2019i commented on sound/soc/sof/loader.c:
```diff
@@ -358,6 +360,68 @@ static int snd_sof_fw_ext_man_parse(struct snd_sof_dev *sdev,
+		memcpy(&module_entry->man4_module_entry, fm_entry, sizeof(*fm_entry));
+
+		if (fm_entry->cfg_count)
+			module_entry->bss_size = fm_config[fm_entry->cfg_offset].is_bytes;
```
"fm_entry->cfg_offset" needs to be validated before use. We have no guarantees that a valid FW (and header) is supplied.

ranj063 replied:
just a quick question @bardliao .  if a module doesnt have a config in the manifest, what is its cfg_offset? If it is 0, wouldnt this be incorrect? you'd be copying the bss_size from another module's config possibly?

bardliao replied:
> just a quick question @bardliao . if a module doesnt have a config in the manifest, what is its cfg_offset? If it is 0, wouldnt this be incorrect? you'd be copying the bss_size from another module's config possibly?

If a module doesnt have a config, its `cfg_count` will be 0. So, it will not pass the `if (fm_entry->cfg_count)` test

ranj063 replied:
@bardliao can cfg_count be > 1?

bardliao replied:
> @bardliao can cfg_count be > 1?

Yes, and the `is_bytes` field will be the same for all of the fm_configs belong to the fm_entry.

kv2019i replied:
@bardliao  You don't really need "fm_config" until this line. So you could move the init of fm_config from L407 to here. For readability, having the space check together with use of the variable, is easier to follow (and verify).

bardliao replied:
> @bardliao You don't really need "fm_config" until this line. So you could move the init of fm_config from L407 to here. For readability, having the space check together with use of the variable, is easier to follow (and verify).

But here is inside a for loop. So I don't think I can move `remaining -= (fw_header->num_module_entries * sizeof(*fm_entry));` here. And it will be less readable if we put `fm_config = (struct sof_man4_module_config *)(fm_entry + fw_header->num_module_entries);` and `remaining -= (fw_header->num_module_entries * sizeof(*fm_entry));` on two places. 
So maybe move those two lines right before the for loop?

kv2019i replied:
You are correct @bardliao , keep fm_config init in the original place.

This segment is difficult to follow as each fm_entry will also access fm_config, so you are consuming data from two places in the binary in this loop. So it's not so clear where to actually modify remaining as you are going back and forth in the blob. And you can do all the checks outside the loop either as you need to parse each fm_entry to see how many fm_config entries is has.

I think the current code is probably the best compromise.

kv2019i commented on sound/soc/sof/loader.c:
```diff
@@ -358,6 +360,68 @@ static int snd_sof_fw_ext_man_parse(struct snd_sof_dev *sdev,
+
+	hdr = (struct sof_ext_manifest4_hdr *)fw->data;
+
+	fw_header = (struct sof_man4_fw_binary_header *)(fw->data + hdr->len +
```
"hdr->len" needs to be validated before use. This must not point beyond the actual FW size in any circumstance.

ujfalusi replied:
_if_ `fw->size == hdr->len + SOF_MAN4_FW_HDR_OFFSET` then there is no space for 'fw_header', no?
Is it mandatory to have 'fw_header'? if it is then the prior check should be extended.

kv2019i replied:
@bardliao I think this is still a valid concern. On L390 you access "fw_header->len" but if fw->size is exactly hrd->len+SOF_MAN4_FW_HDR_OFFSET, you'll end up accessing beyond the buffer.

bardliao replied:
> @bardliao I think this is still a valid concern. On L390 you access "fw_header->len" but if fw->size is exactly hrd->len+SOF_MAN4_FW_HDR_OFFSET, you'll end up accessing beyond the buffer.

Right, that's why I changed the test on L381 from `<` to `<=`. Did I miss something?

kv2019i replied:
@bardliao Hmm, but that's not enough. "fw_header->len" is a 32bit field at offset 4 inside fw_header and the check on L381 just check there's one valid byte at "hdr->len + SOF_MAN4_FW_HDR_OFFSET". I think you should check there is at least "sizeof(*fw_header)" of data in the blob, before you use the casted fw_header pointer.

kv2019i commented on sound/soc/sof/loader.c:
```diff
@@ -358,6 +360,68 @@ static int snd_sof_fw_ext_man_parse(struct snd_sof_dev *sdev,
+
+	fm_entry = (struct sof_man4_module *)((void *)fw_header + fw_header->len);
+	fm_config = (struct sof_man4_module_config *)(fm_entry + fw_header->num_module_entries);
+
```
Same here, "num_module_entries" needs to be validated.

kv2019i reviewed (changes requested):
Some input data validation concerns in the parsing functions.

bardliao commented:
> Some input data validation concerns in the parsing functions.

@kv2019i data validation is added

bardliao commented:
> > Thanks @plbossart. I checked those elements and we don't need most of them.
> 
> Please organize the headers on the SOF repo first: separate in different files the structures shared with the kernel versus not.
> 
> Then, submit to the linux repo an exact copy of the relevant .h file(s). So future maintenance can be achieved with a simple `diff` and patches can be shared between the two repos.
> 
> In other words this review should not happen in the Linux repo, it should happen in the sof repo first.

@marc-hb Agreed. The relevant .h file in SOF repo is https://github.com/thesofproject/rimage/blob/89600974030fcbca903894c327c3c64afcede14d/src/include/rimage/cavs/cavs_ext_manifest.h
That is not exactly the same but I am trying to let them as similar as possible.

RanderWang commented:
> > Thanks @plbossart. I checked those elements and we don't need most of them.
> 
> Please organize the headers on the SOF repo first: separate in different files the structures shared with the kernel versus not.
> 
> Then, submit to the linux repo an exact copy of the relevant .h file(s). So future maintenance can be achieved with a simple `diff` and patches can be shared between the two repos.
> 
> In other words this review should not happen in the Linux repo, it should happen in the sof repo first.

it is not used by fw, just by rimage. And it is already in rimage

marc-hb commented:
> it is not used by fw, just by rimage. 

It is now, you just added that dependency:
https://github.com/thesofproject/sof/commit/cbf5aa6fbc63ef6595c4ca7a2e665d1c73c346ca

> And it is already in rimage

Then the code review for the split between internal structures and interface with the kernel should happen in the rimage repo. Different repo, same logic. Once the external_manifest_interface.h has been reviewed, the file can be simply copied here.

marc-hb commented:
> That is not exactly the same but I am trying to let them as similar as possible.

Is there any reason they cannot be exactly identical eventually?

ujfalusi commented on sound/soc/sof/loader.c:
```diff
@@ -358,6 +360,94 @@ static int snd_sof_fw_ext_man_parse(struct snd_sof_dev *sdev,
+	struct sof_man4_module *fm_entry;
+	int i;
+
+	if (fw->size < sizeof(hdr)) {
```
Probably you meant:
```c
if (fw->size < sizeof(*hdr)) {
```

bardliao replied:
> Probably you meant:
> 
> ```c
> if (fw->size < sizeof(*hdr)) {
> ```

Indeed, thanks

ujfalusi commented on sound/soc/sof/loader.c:
```diff
@@ -358,6 +360,94 @@ static int snd_sof_fw_ext_man_parse(struct snd_sof_dev *sdev,
+	fw_header = (struct sof_man4_fw_binary_header *)(fw->data + hdr->len +
+							 SOF_MAN4_FW_HDR_OFFSET);
+
+	if (fw->size < ((void *)fw_header - (void *)fw->data) + fw_header->len) {
```
This is
```c
if (fw->size < hdr->len + SOF_MAN4_FW_HDR_OFFSET + fw_header->len) {
```
and can be understood.

bardliao replied:
> if fw->size == hdr->len + SOF_MAN4_FW_HDR_OFFSET then there is no space for 'fw_header', no?
Is it mandatory to have 'fw_header'? if it is then the prior check should be extended.

Yes, it is mandatory. I will change the check to `if (fw->size <= hdr->len + SOF_MAN4_FW_HDR_OFFSET)`, thanks.

> This is
if (fw->size < hdr->len + SOF_MAN4_FW_HDR_OFFSET + fw_header->len) {
and can be understood.

Good suggestion, thanks.

ujfalusi commented on sound/soc/sof/loader.c:
```diff
@@ -358,6 +360,94 @@ static int snd_sof_fw_ext_man_parse(struct snd_sof_dev *sdev,
+	fm_entry = (struct sof_man4_module *)((void *)fw_header + fw_header->len);
+
+	if (fw->size < ((void *)fm_entry - (void *)fw->data) +
+			fw_header->num_module_entries * sizeof(*fm_entry)) {
```
It might be again probably just me, but I find this more digestible:
```c
if (fw->size < (hdr->len + SOF_MAN4_FW_HDR_OFFSET + fw_header->len +
		fw_header->num_module_entries * sizeof(*fm_entry)) {
```
and do the `fm_entry = ...` after this check?

ujfalusi replied:
Does each fm_entry must have a matching fm_config? If so then that should be also incorporated into this check?

bardliao replied:
> Does each fm_entry must have a matching fm_config? If so then that should be also incorporated into this check?

No, `fm_entry->cfg_count` can be 0 which means there is no fm_config for this fm_entry.

kv2019i replied:
@bardliao I can see benefits to both approaches, but given the parsing has such many steps, I wonder if this would be easier to read (and maintain) if you'd maintain a relative "remaining = fw->size;" like is done in loader for V3 manifest. Then at each step, you'd verify whether the next parsed element fits the remaining available unparsed FW file.

bardliao commented:
> > That is not exactly the same but I am trying to let them as similar as possible.
> 
> Is there any reason they cannot be exactly identical eventually?

1. We don't want to use bitfields on linux kernel for the compatibility issue.
2. We were not aligned with the naming. For example, ext_manifest4 vs. cavs_ext_manifest.

marc-hb commented:
> We don't want to use bitfields on linux kernel for the compatibility issue.
> We were not aligned with the naming. For example, ext_manifest4 vs. cavs_ext_manifest.

I understand Linux has stricter rules. But why can't we use the solutions and naming required by Linux in the firmware?

bardliao commented:
> Thanks @bardliao and sorry for late review. This looks good. One remaining open I have is whether the parsing and verifying code would be easier to read if using a temporary "remaining" variable. What do you think?

Good suggestion @kv2019i. Fixed it now.

kv2019i commented on sound/soc/sof/loader.c:
```diff
@@ -358,6 +360,98 @@ static int snd_sof_fw_ext_man_parse(struct snd_sof_dev *sdev,
+	dev_dbg(sdev->dev, " fw %s: header length %x, module num %d", fw_header->name,
+		fw_header->len, fw_header->num_module_entries);
+
+	fm_entry = (struct sof_man4_module *)((void *)fw_header + fw_header->len);
```
Hmm, let's avoid void pointer arithmetic. It's an gcc extension and not really needed here. So please use (u8 *) instead.

bardliao replied:
> Hmm, let's avoid void pointer arithmetic. It's an gcc extension and not really needed here. So please use (u8 *) instead.

ack

kv2019i commented on sound/soc/sof/loader.c:
```diff
@@ -358,6 +360,98 @@ static int snd_sof_fw_ext_man_parse(struct snd_sof_dev *sdev,
+	const struct sof_ext_manifest4_hdr *hdr;
+
+	hdr = (struct sof_ext_manifest4_hdr *)fw->data;
+	if (hdr->id == SOF_EXT_MAN4_MAGIC_NUMBER)
```
Here too, we cannot blindly trust fw data is valid, but have to check whether there is enough bytes to parse the header.

bardliao replied:
> Here too, we cannot blindly trust fw data is valid, but have to check whether there is enough bytes to parse the header.

ack

kv2019i commented on sound/soc/sof/loader.c:
```diff
@@ -358,6 +360,101 @@ static int snd_sof_fw_ext_man_parse(struct snd_sof_dev *sdev,
+	dev_dbg(sdev->dev, " fw %s: header length %x, module num %d", fw_header->name,
+		fw_header->len, fw_header->num_module_entries);
+
+	fm_entry = (struct sof_man4_module *)((u8 *)fw_header + fw_header->len);
```
You don't need fm_entry here yet, so this could be before L415. This is really the first place you need fm_entry. This will not impact the sizeof calculations (L401 et al, this just need to know the type).

ujfalusi replied:
@bardliao , can you address @kv2019i's comment?

bardliao replied:
> @bardliao , can you address @kv2019i's comment?

Yes, I moved `fm_entry = (struct sof_man4_module *)((u8 *)fw_header + fw_header->len);` and the `remaining` calculation right before `fm_config = (struct sof_man4_module_config *)(fm_entry + fw_header->num_module_entries);`. Or are you suggesting that moving `fm_entry` only but not the `remaining ` calculation?
Like
```
        remaining -= fw_header->len;

        if (remaining < fw_header->num_module_entries * sizeof(*fm_entry)) {
                dev_err(sdev->dev, "Invalid num_module_entries %d\n",
                        fw_header->num_module_entries);
                return -EINVAL;
        }
        fm_entry = (struct sof_man4_module *)((u8 *)fw_header + fw_header->len);
        fm_config = (struct sof_man4_module_config *)(fm_entry + fw_header->num_module_entries);
``` 
The reason I kept the `remaining ` calculation right after pointer assignments was that I think it is more readable.

kv2019i replied:
I'm good with this. Strictly speaking, you only consume/access the data when the pointer is dereferences (at L415), so in that sense more correctly place to adjust remaining would be L416. But I think this is now close enough and may be easier to read.

kv2019i commented on sound/soc/sof/loader.c:
```diff
@@ -358,6 +360,101 @@ static int snd_sof_fw_ext_man_parse(struct snd_sof_dev *sdev,
+	const struct sof_ext_manifest4_hdr *hdr;
+
+	if (fw->size <= sizeof(*hdr))
+		goto man3;
```
Minor note: not sure how happy upstream is with non-error uses of goto. You could have put this also as:

```
const struct sof_ext_manifest4_hdr *hdr = (struct sof_ext_manifest4_hdr *)fw->data;

if (fw->size > sizeof(*hdr) && hdr->id == SOF_EXT_MAN4_MAGIC_NUMBER) 
      return snd_sof_fw_ext_man4_parse(sdev, fw);

return snd_sof_fw_ext_man3_parse(sdev, fw);
```

bardliao replied:
Thanks for suggestion @kv2019i . Done

kv2019i reviewed (dismissed):
Thanks @bardliao , looks good now! I have one style related issue for snd_sof_fw_ext_man_parse() implementation, see code alternative in the comment. But that's not a blocking concern, so I'm ok to approve.

ujfalusi commented on sound/soc/sof/loader.c:
```diff
@@ -358,6 +360,101 @@ static int snd_sof_fw_ext_man_parse(struct snd_sof_dev *sdev,
+	module_entry = sdev->fw_modules;
+
+	fm_entry = (struct sof_man4_module *)((u8 *)fw_header + fw_header->len);
+	remaining -= fw_header->len;
```
Just out of curiosity: how would this plays out if at this point (or in other places) where you have `a -= (b)` we would have `b > a` ?

ujfalusi replied:
never mind, you have the check a bit further back.

ujfalusi reviewed (dismissed):
@bardliao, looks good, thank you for the updates.

plbossart commented on include/sound/sof/ext_manifest4.h:
```diff
@@ -0,0 +1,118 @@
+#define MAX_MODULE_NAME_LEN     8
+#define MAX_FW_BINARY_NAME      8
+#define DEFAULT_HASH_SHA256_LEN 32
+#define SOF_MAN4_FW_HDR_OFFSET	0x2000
```
a bit of consistency for the define wouldn't hurt, eg. SOF_MAN4 prefix everywhere?

plbossart commented on include/sound/sof/ext_manifest4.h:
```diff
@@ -0,0 +1,118 @@
+	u32 id;
+	u32 len; /* length of extension manifest */
+	u16 version_major;
+	u16 version_minor;
```
possible endianness issue?

bardliao replied:
> possible endianness issue?

Yes, but do we need to support big endian? We will test `hdr->id == SOF_EXT_MAN4_MAGIC_NUMBER` in `snd_sof_fw_ext_man_parse` and it will be false if the firmware binary is built by the different type of endianness.
In other words, we will return error if the firmware binary is built by the different type of endianness.

plbossart commented on include/sound/sof/ext_manifest4.h:
```diff
@@ -0,0 +1,118 @@
+	uint16_t	major_version;
+	uint16_t	minor_version;
+	uint16_t	hotfix_version;
+	uint16_t	build_version;
```
endianness?

kv2019i replied:
@plbossart  Hmm, we don't supported mixed-endian CPU/FW in other interfaces either. E.g. just looking at SOF_ABI, it's parsed the same way. We check magic and then cast the version assuming FW follows CPU endianess. I think we discussed this in the past, but I think this would need a bigger rework to support across the board.

plbossart commented on include/sound/sof/ext_manifest4.h:
```diff
@@ -0,0 +1,118 @@
+	uint16_t	cfg_count;
+	uint32_t	affinity_mask;
+	uint16_t	instance_max_count;
+	uint16_t	instance_stack_size;
```
2 endianness issues here.

plbossart commented on include/sound/sof/ext_manifest4.h:
```diff
@@ -0,0 +1,118 @@
+	uint32_t	affinity_mask;
+	uint16_t	instance_max_count;
+	uint16_t	instance_stack_size;
+	struct sof_man4_segment_desc	segments[3];
```
magic number

plbossart commented on include/sound/sof/ext_manifest4.h:
```diff
@@ -0,0 +1,118 @@
+} __packed;
+
+struct sof_man4_module_config {
+	uint32_t	par[4];		/* module parameters */
```
magic number

plbossart commented on sound/soc/sof/loader.c:
```diff
@@ -358,6 +360,96 @@ static int snd_sof_fw_ext_man_parse(struct snd_sof_dev *sdev,
+{
+	const struct sof_ext_manifest4_hdr *hdr = (struct sof_ext_manifest4_hdr *)fw->data;
+
+	if (fw->size > sizeof(*hdr) && hdr->id == SOF_EXT_MAN4_MAGIC_NUMBER)
```
this logical combination is far from obvious to review. A comment wouldn't hurt.

bardliao commented:
@plbossart I addressed all comments except the endianness issue. I need more discussion on the endianness issue.

plbossart commented:
> @plbossart I addressed all comments except the endianness issue. I need more discussion on the endianness issue.

Sorry I wasn't asking for a change, it would be fine to just have a comment that this assumes little-endian order for compatibility with Windows firmware. That way upstream reviewers will know it's by-design and not a mistake.

bardliao commented:
> > @plbossart I addressed all comments except the endianness issue. I need more discussion on the endianness issue.
> 
> Sorry I wasn't asking for a change, it would be fine to just have a comment that this assumes little-endian order for compatibility with Windows firmware. That way upstream reviewers will know it's by-design and not a mistake.

Oh, I see. Adding a sentence `Note that not all items in the struct are 32 bits length, and we assume the firmware is little-endian.` at the beginning of ext_manifest4.h now.

ranj063 commented on include/sound/sof/ext_manifest4.h:
```diff
@@ -0,0 +1,122 @@
+#define SOF_MAN4_DEFAULT_HASH_SHA256_LEN 32
+#define SOF_MAN4_SEGMENT_NUM 3
+#define SOF_MAN4_MODULE_PARAMS_NUM 4
+#define SOF_MAN4_FW_HDR_OFFSET 0x2000
```
@bardliao the FW_HDR_OFFSET is not the same for cavs1.5 abd cavs1.8. Can we please abstract this?

bardliao replied:
> @bardliao the FW_HDR_OFFSET is not the same for cavs1.5 abd cavs1.8. Can we please abstract this?

Thanks @ranj063. I did some change, could you review it? Thanks.

ranj063 commented on include/sound/sof/ext_manifest4.h:
```diff
@@ -0,0 +1,122 @@
+#define SOF_MAN4_DEFAULT_HASH_SHA256_LEN 32
+#define SOF_MAN4_SEGMENT_NUM 3
+#define SOF_MAN4_MODULE_PARAMS_NUM 4
+#define SOF_MAN4_FW_HDR_OFFSET 0x2000
```
@bardliao thanks but you didnt use the struct defined in the first patch?

bardliao replied:
> @bardliao thanks but you didnt use the struct defined in the first patch?

Yes, I did. See [snd_sof_fw_ext_man4_parse](https://github.com/thesofproject/linux/pull/3176/commits/41951dfd31585a5ca5d216bedc8d89ab99ab7555#diff-a4132048dde4b0a1f8b761545f01a09bb681f6ecdf49d0d7bb374084ae4843aaR363)
```
static int snd_sof_fw_ext_man4_parse(struct snd_sof_dev *sdev, const struct firmware *fw)
{
	struct snd_sof_pdata *sof_pdata = sdev->pdata;
	const struct sof_dev_desc *desc = sof_pdata->desc;
	const struct snd_sof_ext_man_pdata *ext_man_pdata = desc->ext_man_pdata;
...
	if (ext_man_pdata)
		fw_hdr_offset = ext_man_pdata->fw_hdr_offset;
	else
		fw_hdr_offset = SOF_MAN4_FW_HDR_OFFSET;
...
}
```
